File: 014_longest_common_prefix.js
var longestCommonPrefix = function(strs) {
    if (strs.length == 0 || strs[0] == "") { return "" }
    if (strs.length == 1) { return strs[0] }

    let lcp = ""

    for (let i = 0; i < strs[0].length; i++){
        let hold = strs[0][i]

        for (let j = 0; j < strs.length; j++){
            if (strs[j][i] == undefined || strs[j][i] != hold[0]) { return lcp }
        }
        lcp += hold
    }
    return lcp
};

let testString1 = ["flower","flow","flight"]
let testStrings2 = ["dog","racecar","car"]

console.assert(longestCommonPrefix(testString1) === "fl", 1);
console.assert(longestCommonPrefix(testStrings2) === "", 2);
console.assert(longestCommonPrefix([""]) === "", 3);
console.assert(longestCommonPrefix(["a"]) === "a", 4);
console.assert(longestCommonPrefix(["flower","flower","flower","flower"]) === "flower", 5);

File: 014_longest_common_prefix.pl
use warnings;

sub longestCommonPrefix{
    my @strs = @_;

    return "" if ( scalar @strs == 0 || $strs[0] eq "" );
    return $strs[0] if (scalar @strs == 1);

    my $lcp = "";

    for (my $i = 0; $i < length $strs[0]; $i++){
        my $hold = substr($strs[0], $i, 1);

        for (my $j = 0; $j < scalar @strs; $j++){
            return $lcp if (!defined substr($strs[$j], $i, 1) or substr($strs[$j], $i, 1) ne $hold);
        }
        $lcp = $lcp.$hold;
    }
    return $lcp;
}


my @testString1 = ("flower","flow","flight");
my @testStrings2 = ("dog","racecar","car");


longestCommonPrefix(@testString1) eq "fl" ? print "PASSED\n" : print "FAILED\n";
longestCommonPrefix(@testString2) eq "" ? print "PASSED\n" : print "FAILED\n";
longestCommonPrefix(("")) eq "" ? print "PASSED\n" : print "FAILED\n";
longestCommonPrefix(("a")) eq "a" ? print "PASSED\n" : print "FAILED\n";
longestCommonPrefix((("flower","flower","flower","flower"))) eq "flower" ? print "PASSED\n" : print "FAILED\n";